lib/migrate_training.rb

```diff
@@ -0,0 +1,154 @@
+  class Error < StandardError
+  end
+
+  extend Dry::Initializer
+
+  # @return [Boolean]
+  option :truncate, Types::Bool, default: proc { false }, reader: :private
+  # @return [Boolean]
+  option :simulate, Types::Bool, default: proc { false }, reader: :private
+  # @return [Boolean]
+  option :verbose, Types::Bool, default: proc { true }, reader: :private
+  # @return [Integer, nil]
+  option :resume, Types::Integer, optional: true, reader: :private
+
+  # @return [?]
+  def call
+    ActiveRecord::Base.transaction do
+      log "Migration started: #{Time.zone.now}"
+      truncate! if truncate
+      migrate!
+      log "Migration finished: #{Time.zone.now}"
+      raise Error if simulate
+    end
+  end
+
+private
+
+  # @return [PG::Result]
+  def truncate!
+    log 'Truncate responses and assessments'
+    ActiveRecord::Base.connection.execute 'TRUNCATE responses, assessments RESTART IDENTITY'
+  end
+
+  # @return [nil]
+  def migrate!
+    UserAnswer.find_each(start: resume) do |user_answer|
+      if valid?(user_answer)
+        response = process_user_answer(user_answer)
+        log response.attributes.to_json, alert: false
+      else
+        log "User: #{user_answer.user_id} UserAnswer: #{user_answer.id}"
+      end
+    end
+  end
+
+  # @param user_answer [UserAnswer]
+  # @return [Boolean]
+  def valid?(user_answer)
+    if user_answer.name.blank? ||
+        user_answer.module.blank? ||
+        user_answer.assessments_type.blank? ||
+        user_answer.question.nil? ||
+        !user_answer.question.respond_to?(:question_type)
+      false
+    else
+      true
+    end
+  end
+
+  # @param user_answer [UserAnswer]
+  # @return [Response]
+  def process_user_answer(user_answer)
+    assessment = process_user_assessment(user_answer)
+    params = response_params(user_answer)
+
+    Response.create(**params, assessment_id: assessment&.id)
+  end
+
+  # @param user_answer [UserAnswer]
+  # @return [Assessment, nil]
+  def process_user_assessment(user_answer)
+    return unless user_answer.question.summative_question?
+
+    if user_answer.user_assessment_id.nil?
+      assessment =
+        Assessment
+          .create_with(started_at: user_answer.created_at)
+          .find_or_create_by(user_id: user_answer.user_id, training_module: user_answer.module)
+    else
+      user_assessment = UserAssessment.find(user_answer.user_assessment_id)
+      params = assessment_params(user_assessment)
+      assessment = Assessment.find_or_create_by(params)
+
+      if assessment.score.nil?
+        if user_assessment.score.to_i > 100
+          assessment.update(score: 100)
+        else
+          assessment.update(score: user_assessment.score)
+        end
+      end
+    end
+
+    assessment
+  end
+
+  # @param user_answer [UserAssessment]
+  # @return [Hash<Symbol=>Mixed>]
+  def assessment_params(user_assessment)
+    {
+      id: user_assessment.id,                                                          # int primary key
```

Comment:
@peterdavidhamilton Are we better off letting the db generate the id instead of manually setting it here?

Reply:
@jack-coggin your DB contents have likely changed since commenting but running it with `truncate` enabled would resolve it. Existing records in the new tables need to be cleared out first.